### PR TITLE
[Kernels] Support blocked fp8 quantization for compressed tensors MoE

### PR DESCRIPTION
**Pull request by bnellnm**
## Purpose
Add fp8 blocked quantization support to compressed tensors MoE.

## Test Plan
Run lm-eval on the following models:
- RedHatAI/Qwen3-30B-A3B-FP8-BLOCK
- nm-testing/Llama-4-Maverick-17B-128E-Instruct-FP8-BLOCK                                                                                                                                                      
- nm-testing/Llama-4-Scout-17B-16E-Instruct-FP8-BLOCK                                                                                                                                                                               

## Test Result
- RedHatAI/Qwen3-30B-A3B-FP8-BLOCK
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.86|±  |0.0349|
|     |       |strict-match    |     5|exact_match|↑  | 0.95|±  |0.0219|
```

- nm-testing/Llama-4-Maverick-17B-128E-Instruct-FP8-BLOCK                                                                                                                                                        
```
OOM'd on 8 gpu machine.
```

- nm-testing/Llama-4-Scout-17B-16E-Instruct-FP8-BLOCK   
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.94|±  |0.0239|
|     |       |strict-match    |     5|exact_match|↑  | 0.92|±  |0.0273|
```

### CODE REVIEW

**Comment by bnellnm** [comment]
@mgoin , @tlrmchlsmth I think this is ready to go as a first cut.

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 1]
```diff
@@ -602,6 +642,29 @@ def create_weights(self, layer: torch.nn.Module, num_experts: int,
...
+                dtype=torch.float32),
+                                                 requires_grad=False)
+            layer.register_parameter("w2_weight_scale", w2_weight_scale)
+            # Add PER-CHANNEL quantization for FusedMoE.weight_loader.
+            extra_weight_attrs.update(
+                {"quant_method": FusedMoeWeightScaleSupported.BLOCK.value})
+            set_weight_attrs(w13_weight_scale, extra_weight_attrs)
+            set_weight_attrs(w2_weight_scale, extra_weight_attrs)
+            layer.register_parameter("w13_weight_scale_inv", w13_weight_scale)
+            layer.register_parameter("w2_weight_scale_inv", w2_weight_scale)
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Do we need to use _inv here? It seems to get jumbled later on in process_weights_after_loading

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 2]
```diff
@@ -45,10 +48,17 @@
...
+from vllm.utils.deep_gemm import is_deep_gemm_e8m0_used
 
 logger = init_logger(__name__)
 
 
+def _is_col_major(x: torch.Tensor) -> bool:
+    assert x.dim() == 3
+    b, m, n = x.shape
+    return x.stride(0) == m * n and x.stride(1) == 1 and x.stride(2) == m
+
```
Comment (yewentao256) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Let's reuse the code `_is_col_major` in fp8 or make it a util.

**Review by yewentao256 — COMMENTED** [review]
Looks good and thanks for the work!
Could you also add performance metric using `vllm bench throughput...`?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 1 above)
Comment (bnellnm) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: The deep gemm warmup code still looks for `_inv`.  I hacked it up to tolerate either.

**Comment by bnellnm** [comment]
Benchmark with DP=4 + deepep_high_throughput
```
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  60.33
Total input tokens:                      1021533
Total generated tokens:                  110757
Request throughput (req/s):              16.58
Output token throughput (tok/s):         1835.83
Peak output token throughput (tok/s):    4559.00
Peak concurrent requests:                1000.00
Total Token throughput (tok/s):          18768.05
---------------Time to First Token----------------
Mean TTFT (ms):                          20646.11
Median TTFT (ms):                        23193.82
P99 TTFT (ms):                           50440.13
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          176.28
Median TPOT (ms):                        183.99
P99 TPOT (ms):                           310.81
---------------Inter-token Latency----------------
Mean ITL (ms):                           173.92
Median ITL (ms):                         114.73
P99 ITL (ms):                            2261.17
==================================================
```

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 3]
```diff
@@ -505,10 +510,13 @@ def __init__(
             self.weight_quant.strategy == QuantizationStrategy.CHANNEL
             and self.input_quant.strategy == QuantizationStrategy.TOKEN)
         if not (per_tensor or per_channel):
-            raise ValueError(
-                "For FP8 Fused MoE layers, we require per tensor "
-                "or channelwise, dynamic per token quantization. Found "
-                f"{self.weight_quant}, {self.input_quant}")
+            assert self.weight_quant.strategy == QuantizationStrategy.BLOCK
+            logger.debug("WQ = %s", str(self.weight_quant))
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: debug cruft?

**Review by mgoin — APPROVED** [review]
